ledger/participant-integration-api/src/test/suite/scala/platform/index/BuffersUpdaterSpec.scala

```diff
@@ -0,0 +1,303 @@
+          Source(
+            immutable.Iterable(
+              offset1 -> update1,
+              offset2 -> update2,
+              offset3 -> update2, /* Wrong on purpose */
+            )
+          )
+            .map(x =>
+              if (x._1._2 == 3L) {
+                println(s"been here $x")
+                throw new RuntimeException("some transient failure")
+              } else x
+            )
+        case (_, 2L) =>
+          Source
+            .single(offset3 -> update3)
+            // Keep alive so we don't trigger unwanted restarts (for the purpose of this test)
+            .concat(Source.never)
+        case unexpected => fail(s"Unexpected re-subscription point $unexpected")
+      }
+
+      val transactionsBufferMock =
+        scala.collection.mutable.ArrayBuffer.empty[(Offset, TransactionLogUpdate)]
+      val updateTransactionsBufferMock = (offset: Offset, update: TransactionLogUpdate) => {
+        transactionsBufferMock += (offset -> update)
+        ()
+      }
+
+      val buffersUpdater = new BuffersUpdater(
+        subscribeToTransactionLogUpdates = sourceSubscriptionFixture,
+        updateTransactionsBuffer = updateTransactionsBufferMock,
+        toContractStateEvents = Map.empty.withDefaultValue(Iterator.empty),
+        updateMutableCache = _ => (),
+        minBackoffStreamRestart = 1.millis,
+        _ => fail("should not be triggered"),
+      )(materializer, loggingContext, scala.concurrent.ExecutionContext.global)
+
+      eventually {
+        transactionsBufferMock should contain theSameElementsAs Seq(
+          offset1._1 -> update1,
+          offset2._1 -> update2,
+          offset3._1 -> update3,
+        )
+
+        buffersUpdater.updaterIndex.get shouldBe offset3
+      }
+    }
+
+    "shutdown is triggered with status code 1 if a non-recoverable error occurs in the buffers updating logic" in {
+      val update1, update2 = transactionLogUpdateMock()
+      val Seq(offset1, offset2) =
+        (1L to 2L).map(idx => Offset.fromByteArray(BigInt(1337L + idx).toByteArray) -> idx)
+
+      val sourceSubscriptionFixture = (_: Offset, _: EventSequentialId) =>
+        Source(immutable.Iterable(offset1 -> update1, offset2 -> update2))
+      val updateTransactionsBufferMock = (offset: Offset, _: TransactionLogUpdate) =>
+        if (offset == offset2._1) throw new RuntimeException("Unrecoverable exception")
+        else ()
+
+      val shutdownCodeCapture = new AtomicInteger(Integer.MIN_VALUE)
+
+      new BuffersUpdater(
+        subscribeToTransactionLogUpdates = sourceSubscriptionFixture,
+        updateTransactionsBuffer = updateTransactionsBufferMock,
+        toContractStateEvents = Map.empty,
+        updateMutableCache = _ => (),
+        minBackoffStreamRestart = 1.millis,
+        shutdownCodeCapture.set,
+      )(materializer, loggingContext, scala.concurrent.ExecutionContext.global)
+
+      eventually {
+        shutdownCodeCapture.get() shouldBe 1
```

Comment:
I assume the exception and/or the shutdown is also logged at ERROR level, is it?

Reply:
Could / should the test assert that as well?

Reply:
Asserted the logging 👍